### DISCUSSION TITLE
Typeclass vs Trait vs Interface/聊聊多态 | XX’s Blog

### DISCUSSION THREAD

**Discussion by xxchan** [xxchan/xxchan.github.io#22] [Announcements]
https://xxchan.github.io/cs/2020/08/17/polymorphism.html 

 Disclamer：由于本人才疏学浅，并且本文涉及的话题内容过多，肯定有不全面甚至错误。本文既不学术也不实践，仅为个人阶段的学习理解，写的也比较随意。如有错误欢迎指正。

**Comment by KaiyuanLiu0** [comment]
**好难，没学过Haskell**

**Comment by xxchan** [comment]
> @LucerneLaa
> **好难，没学过Haskell**

快学

**Comment by vickiegpt** [comment]
怕不是在摘抄）

**Comment by xxchan** [comment]
> @victoryang00
> 怕不是在摘抄）

别骂了😭

**Comment by xxchan** [comment]
终于懂了，row polymorphism 原来就是 structural subtyping

**Comment by im-rinnin** [comment]
推荐两个资料
types and Programming Languages 老生常谈了，但是说到pl还是要看这本吧，建议配合ai一起看，可以跳过证明和实现部分，重要是构建最基本的知识结构，写的太数学太形式化了我反正看不进去
Traits: Composable Units of Behaviour   trait 的设计，大概就是通过去掉状态（成员变量），只保留函数，进行组合，同时组合是没有层级顺序关系的，从而实现非常自由的组合
